File: clients/python/avro_producer.py
"""Iggy producer that serializes messages using Apache Avro.

The Avro schema is defined in ``schemas.py`` (and the canonical source file is
``schemas/event.avsc``).  Messages are sent as raw schemaless Avro bytes so
that the consumer can decode them with the same schema.

Usage::

    python avro_producer.py
"""
import asyncio
from datetime import datetime, timezone

from loguru import logger
from iggy_py import IggyClient, SendMessage

from schemas import avro_serialize

STREAM_NAME = "demo-stream"
TOPIC_NAME = "avro-topic"
PARTITION_ID = 1
SEND_INTERVAL_SECS = 1.0


async def main() -> None:
    client = IggyClient()
    logger.info("Connecting to Iggy server...")
    await client.connect()
    logger.info("Connected. Logging in...")
    await client.login_user("iggy", "iggy")
    logger.info("Logged in as iggy.")
    await init_system(client)
    await produce_messages(client)


async def init_system(client: IggyClient) -> None:
    """Idempotently create the stream and topic if they don't exist."""
    try:
        stream = await client.get_stream(STREAM_NAME)
        if stream is None:
            await client.create_stream(name=STREAM_NAME)
            logger.info(f"Stream '{STREAM_NAME}' created.")
        else:
            logger.info(f"Stream '{STREAM_NAME}' already exists (id={stream.id}).")
    except Exception as e:
        logger.error(f"Error setting up stream: {e}")
        raise

    try:
        topic = await client.get_topic(STREAM_NAME, TOPIC_NAME)
        if topic is None:
            await client.create_topic(STREAM_NAME, TOPIC_NAME, 1)
            logger.info(f"Topic '{TOPIC_NAME}' created.")
        else:
            logger.info(f"Topic '{TOPIC_NAME}' already exists (id={topic.id}).")
    except Exception as e:
        logger.error(f"Error setting up topic: {e}")
        raise


async def produce_messages(client: IggyClient) -> None:
    logger.info(
        f"Producing Avro messages to stream='{STREAM_NAME}' topic='{TOPIC_NAME}' "
        f"partition={PARTITION_ID} every {SEND_INTERVAL_SECS}s. Press Ctrl+C to stop."
    )
    message_id = 0
    while True:
        message_id += 1
        record = {
            "id": message_id,
            "text": "hello from avro producer",
            "ts": datetime.now(timezone.utc).isoformat(),
        }
        payload_bytes = avro_serialize(record)
        message = SendMessage(payload_bytes)
        try:
            await client.send_messages(
                STREAM_NAME,
                TOPIC_NAME,
                PARTITION_ID,
                [message],
            )
            logger.info(f"Sent Avro message #{message_id}: {record}")
        except Exception as e:
            logger.error(f"Failed to send Avro message #{message_id}: {e}")
        await asyncio.sleep(SEND_INTERVAL_SECS)


if __name__ == "__main__":
    asyncio.run(main())

File: clients/rust/src/avro_producer.rs
//! Iggy producer that serializes messages using Apache Avro.
//!
//! The Avro schema is defined in `schemas.rs` (canonical source:
//! `schemas/event.avsc`).  Messages are sent as raw schemaless Avro bytes so
//! that the consumer can decode them with the same schema.
//!
//! Usage:
//!   cargo run --bin avro_producer

#[path = "schemas.rs"]
mod schemas;

use iggy::prelude::*;
use std::error::Error;
use std::str::FromStr;
use std::time::Duration;
use tracing::{error, info};

const STREAM_NAME: &str = "demo-stream";
const TOPIC_NAME: &str = "avro-topic";
const PARTITION_ID: u32 = 1;
const SEND_INTERVAL_SECS: u64 = 1;

#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    tracing_subscriber::fmt::init();

    info!("Connecting to Iggy server...");
    let client = IggyClientBuilder::new()
        .with_tcp()
        .with_server_address("127.0.0.1:8090".to_string())
        .build()?;

    client.connect().await?;
    info!("Connected. Logging in...");
    client.login_user("iggy", "iggy").await?;
    info!("Logged in as iggy.");

    init_system(&client).await?;
    produce_messages(&client).await?;

    Ok(())
}

async fn init_system(client: &IggyClient) -> Result<(), Box<dyn Error>> {
    let stream_id = Identifier::from_str(STREAM_NAME)?;
    let topic_id = Identifier::from_str(TOPIC_NAME)?;

    match client.get_stream(&stream_id).await? {
        Some(stream) => info!("Stream '{}' already exists (id={}).", STREAM_NAME, stream.id),
        None => {
            client.create_stream(STREAM_NAME).await?;
            info!("Stream '{}' created.", STREAM_NAME);
        }
    }

    match client.get_topic(&stream_id, &topic_id).await? {
        Some(topic) => info!("Topic '{}' already exists (id={}).", TOPIC_NAME, topic.id),
        None => {
            client
                .create_topic(
                    &stream_id,
                    TOPIC_NAME,
                    1,
                    Default::default(),
                    None,
                    IggyExpiry::NeverExpire,
                    MaxTopicSize::ServerDefault,
                )
                .await?;
            info!("Topic '{}' created.", TOPIC_NAME);
        }
    }

    Ok(())
}

async fn produce_messages(client: &IggyClient) -> Result<(), Box<dyn Error>> {
    info!(
        "Producing Avro messages to stream='{}' topic='{}' partition={} every {}s. Press Ctrl+C to stop.",
        STREAM_NAME, TOPIC_NAME, PARTITION_ID, SEND_INTERVAL_SECS
    );

    let stream_id = Identifier::from_str(STREAM_NAME)?;
    let topic_id = Identifier::from_str(TOPIC_NAME)?;
    let partitioning = Partitioning::partition_id(PARTITION_ID);

    let mut message_id: i32 = 0;
    loop {
        message_id += 1;
        let ts = chrono::Utc::now().to_rfc3339();

        match schemas::avro_serialize(message_id, "hello from Rust Avro producer", &ts) {
            Ok(avro_bytes) => {
                let mut messages = vec![IggyMessage::builder()
                    .payload(avro_bytes.into())
                    .build()?];

                match client
                    .send_messages(&stream_id, &topic_id, &partitioning, &mut messages)
                    .await
                {
                    Ok(_) => info!(
                        "Sent Avro message #{}: id={} text='hello from Rust Avro producer' ts={}",
                        message_id, message_id, ts
                    ),
                    Err(e) => error!("Failed to send Avro message #{}: {}", message_id, e),
                }
            }
            Err(e) => error!("Failed to serialize Avro message #{}: {}", message_id, e),
        }

        tokio::time::sleep(Duration::from_secs(SEND_INTERVAL_SECS)).await;
    }
}
